### DISCUSSION TITLE
Suggest Workout: LLM-Powered Workout Planning (design)

### DISCUSSION THREAD

**Discussion by dustinmays** [aptx-health/ripit-fitness#868] [Ideas]
## ⚠ Wave/status maps in this doc are frozen historical narrative — do not trust them for current state

**Canonical milestone status** lives in `milestone-visualizer` and the checked-in dependency graph:

- Live status: `msv status aptx-health/ripit-fitness -m "Suggest Workout (LLM-powered)"`
- Ready-to-pick-up: `msv ready`
- Dependency graph: `milestones/suggest-workout.md` (locally, gitignored)
- Lints (PR/issue mismatch, cycles, orphans): `msv doctor`

This document is now scoped to **rationale and locked design decisions** only. Everything below is preserved for historical context but may reference incorrect issue numbers (an early parallel-issue-creation batch shuffled the numbering, and status updates were written from a mental model that never got reconciled against `gh issue list`). Ground truth for issue → scope mapping is always the live GitHub state.

### Postmortem: why the discussion drifted from reality

1. Issues were created in parallel `bash &` batches. GitHub assigns issue numbers in API-completion order, not command order. The stdout log I trusted did not reflect the true assignment.
2. I ran subsequent updates (bodies, comments, labels) targeting numbers from my own log rather than verifying titles.
3. When PRs came in with `Fixes #N` references from the autopilot agent (which correctly used branch → issue mapping), I "corrected" them against my shuffled mental map — introducing the actual bug.
4. The reviewer caught it by reconciling three sources: PR branch name, PR diff, and `gh issue list --milestone`.

### Rules baked in going forward

- Before writing to any issue by number, run `gh issue view <N> --json title` and verify scope.
- Before editing `Fixes #N` on a PR, reconcile branch name + diff + issue body. If any two disagree, flag and stop.
- Wave/status maps live in `msv`, not in prose.
- `msv doctor` is the linter — should run in CI on PR opens (see `docs/msv-notes` — pending).

---

*(Historical content below preserved unedited for context.)*
## Vision

User taps **Suggest Workout**, optionally provides ephemeral context (time available, "deprioritize legs — sore", "feeling strong"), and gets back a workout — an ordered list of exercises (no sets/reps in v1) — that respects their durable training profile, recent training history, equipment, and stated goals. They edit freely with existing tools and log it. Edits feed back into the learning loop.

## The Core Insight: Context Engineering, Not ML

We are not training a model. We are **designing what an LLM sees at inference time** so that a competent off-the-shelf model can produce expert-quality suggestions for *this user, right now*.

> Stop trying to make the LLM smarter; focus on making its input better.

The model picks the exercises. The engineering decides what the model knows.

## Why LLM Is the Right Tool Here

| Deterministic code does | LLM does |
|---|---|
| Snapshot of recent volume, FAU deficits, days-since-trained | Interpret user goals, reconcile conflicts |
| Exercise catalog filtering (equipment, banned list) | Pick the actual exercises and order |
| Final validation (no dupes, time budget respected, IDs exist) | Generate "why" copy that builds user trust |
| Push to workout creation API | — |

Trade-offs:
- ✅ Inputs are fuzzy ("upper-body tilt, spare legs for biking, improve bench")
- ✅ Constraints conflict and need prioritization
- ✅ Explanation matters as much as selection ("went lighter on legs because you bike Saturday")
- ❌ Latency >100ms — acceptable for an explicit "Suggest" tap, not for typeahead
- ❌ Probabilistic — needs deterministic shell for validation/retry

## Architecture

```
Request (user taps Suggest, provides ephemeral context)
  ↓
Build candidate list (deterministic: equipment filter, FAU relevance, not-banned)
  ↓
Build training-state payload (deterministic: EWMA calibration, recent feedback, FAU recency)
  ↓
Build prompt (durable profile + ephemeral context + training state + candidates)
  ↓
LLM call (structured output: exercise_ids[] + per-exercise rationale + summary)
  ↓
Validate (zod): IDs exist in candidate list, count fits time budget, no dupes
  ↓
On failure: one retry with error feedback in prompt
  ↓
Transform to workout creation payload
  ↓
POST to existing freestyle/adhoc workout API
  ↓
Return to user with rationale text + link to start workout
```

## Durable vs Ephemeral Data

**Durable** (`UserTrainingProfile`, edited rarely):
- Goal sentences (free text — LLM interprets)
- Equipment available
- Hard rules: banned exercises (injuries), ratio targets, weekly intent ("1 heavy leg day")

**Ephemeral** (per request, not persisted beyond the resulting workout):
- "I have 40 minutes"
- "Deprioritize glutes — sore"
- "Feeling strong"
- "Biking tomorrow"

Boundary chosen so adding new ephemeral signals = new request field, no migration. Durable profile stays small and stable.

## The Calibration Story (Lightweight Learning)

"Heavy" varies per user. We learn it without building an ML pipeline.

**EWMA per movement pattern**:
```
estimate = α * new_observation + (1 - α) * previous_estimate
```
Track top working weight per `movement_pattern` (horizontal_push, hip_hinge, etc.) as an EWMA. LLM gets `current_estimate` + `recent_observations` and interprets ("heavy = ~10% above current estimate"). No threshold stored.

**Beta distributions for preferences**:
Store `(likes_count, dislikes_count)` per `(user, exercise)`. Mean of `Beta(α+1, β+1)` = preference score; spread = confidence. Handles cold start (Beta(1,1) = uniform = no opinion). Three sessions of data already produces a usable signal — unlike a raw average.

**Cross-user pooling**: deferred. Worth doing when we have 20+ users.

## Implicit Feedback (the strongest signal)

When a workout originated from a suggestion (`Workout.sourceSuggestionId`):
- **Swaps** (X → Y): X.dislikes += 1, Y.likes += 1. Strongest signal — user wanted *this kind* of exercise but preferred Y.
- **Additions**: "AI underweighted this FAU." Biases future ratio interpretation.
- **Deletions**: "AI overweighted this" or "ran out of time." Weaker; repeated deletion of same exercise approaches soft ban.
- **Reorders**: ignored in v1 (mostly noise).

Zero new UI. Existing edit handlers call `recordImplicitFeedback()` when the workout has a `sourceSuggestionId`. Learning happens automatically as the user uses the app normally.

## Exercise Tagging — LLM-as-Bootstrap

We need `movement_pattern` (compound/isolation, push/pull/hinge/squat/carry/etc.) and `intensity_class` (heavy/moderate/light) on `ExerciseDefinition`. Instead of a manual tagging sprint:

- On first need, LLM classifies the exercise; result cached to DB
- One-time backfill script pre-tags top 100 exercises (~$0.10)
- New exercises tagged lazily on first use
- Costs trend to zero as catalog gets covered

## Provider-Agnostic LLM Client

`LLM_PROVIDER_URL`, `LLM_API_KEY`, `LLM_MODEL` env vars. Use `openai` SDK pointed at any OpenAI-compatible endpoint (DeepInfra, Together, Groq, local Ollama, Anthropic via proxy). Single env swap to change providers.

- Structured output via JSON schema mode
- zod validation + one retry on validation failure
- Per-provider quirks isolated in `lib/llm/client.ts`

Cost analysis: ~80–150 candidates after filtering × ~30 tokens each = 2.4k–4.5k tokens catalog context. Pennies per request on Haiku-tier or Llama 3.1 8B.

## Onboarding Agent

Conversational LLM interview at signup (re-runnable from Settings) writes the `UserTrainingProfile`. 6–10 questions, chip-style multi-choice where helpful. Solves blank-slate problem and produces better settings than any form could.

For less-knowledgeable users: agent guides with general questions ("what brought you to lifting?") and translates answers into structured profile fields.

For knowledgeable users: agent accepts complex goal sentences directly ("upper tilt, spare legs for biking, improve bench") and confirms its interpretation.

## What We Chose NOT to Do

- **Manual movement-pattern tagging sprint**: LLM-as-bootstrap is cheaper and handles new exercises automatically
- **Per-user threshold storage** ("user's heavy bench = 185 lbs"): EWMA over observations is more honest and adapts automatically
- **Neural anything**: tiny per-user data + LLM priors >> any tiny neural model
- **Contextual bandits**: real ML technique that fits this problem, but overkill until many more users exist
- **Cross-user pooling**: deferred until 20+ user base
- **Always-on post-workout RPE prompt**: friction tax too high. Implicit feedback from edits is the v1 signal; explicit 😴/👍/🔥 buttons are a v2 enhancement, asked only when AI is least sure

## Milestone Breakdown

Issues live in [Milestone 15](https://github.com/aptx-health/ripit-fitness/milestone/15).

**Foundation (autopilot)**:
1. `UserTrainingProfile` schema
2. `UserExercisePreference` table (Beta storage)
3. `Workout.sourceSuggestionId` + `Suggestion` table
4. LLM client wrapper
5. LLM-as-bootstrap exercise tagger + backfill
6. Learning helpers (EWMA + Beta math)
7. Training-state builder
8. Implicit feedback hooks on edit endpoints

**Headline features (interactive design)**:
9. Onboarding interview agent
10. Suggest Workout v1

**Polish/orthogonal (ux-autopilot)**:
11. Equipment checklist UI
12. FAU aggregation cleanup (days/weeks, zero-volume semantics)
13. Ratio presets

**Docs**:
14. `docs/LEARNING_NOTES.md` — design rationale + future ML directions

## Future Directions

When the user base or per-user data grows, the foundation here unlocks:
- **Cross-user pooling**: "users with your goals tolerate 40% volume reduction on light days"
- **Contextual bandits**: principled exploration/exploitation in suggestions
- **Per-user model fine-tuning**: if it ever makes sense, the structured logs are already there
- **Coach-grade explanations**: richer rationale UI showing how each input shaped the suggestion
- **Auto-program generation**: same machinery, longer time horizon — generate a multi-week loose program from goals + history
